crates/oxc_linter/src/service/runtime.rs

```diff
@@ -2,43 +2,121 @@ use std::{
     borrow::Cow,
     ffi::OsStr,
     fs,
+    mem::take,
     path::{Path, PathBuf},
     rc::Rc,
-    sync::Arc,
+    sync::{Arc, mpsc},
 };
 
-use rayon::{
-    iter::{IntoParallelRefIterator, ParallelBridge},
-    prelude::ParallelIterator,
-};
-use rustc_hash::FxHashSet;
+use indexmap::IndexSet;
+use rayon::iter::ParallelDrainRange;
+use rayon::{Scope, iter::IntoParallelRefIterator, prelude::ParallelIterator};
+use rustc_hash::{FxBuildHasher, FxHashMap, FxHashSet};
+use self_cell::self_cell;
+use smallvec::SmallVec;
 
 use oxc_allocator::Allocator;
 use oxc_diagnostics::{DiagnosticSender, DiagnosticService, Error, OxcDiagnostic};
 use oxc_parser::{ParseOptions, Parser};
 use oxc_resolver::Resolver;
-use oxc_semantic::SemanticBuilder;
-use oxc_span::{SourceType, VALID_EXTENSIONS};
+use oxc_semantic::{Semantic, SemanticBuilder};
+use oxc_span::{CompactStr, SourceType, VALID_EXTENSIONS};
 
+use super::LintServiceOptions;
 use crate::{
     Fixer, Linter, Message,
     loader::{JavaScriptSource, LINT_PARTIAL_LOADER_EXT, PartialLoader},
     module_record::ModuleRecord,
     utils::read_to_string,
 };
 
-use super::{
-    LintServiceOptions,
-    module_cache::{ModuleCache, ModuleState},
-};
-
 pub struct Runtime {
     cwd: Box<Path>,
     /// All paths to lint
-    paths: FxHashSet<Box<Path>>,
+    paths: IndexSet<Arc<OsStr>, FxBuildHasher>,
     pub(super) linter: Linter,
     resolver: Option<Resolver>,
-    modules: ModuleCache,
+
+    #[cfg(test)]
+    pub(super) test_source: std::sync::RwLock<Option<String>>,
+}
+
+/// Output of `Runtime::process_path`
+struct ModuleProcessOutput {
+    /// All paths in `Runtime` are store as `OsStr`, because `OsStr` hash is faster
+    /// than `Path` - go checkout their source code.
+    path: Arc<OsStr>,
+    processed_module: ProcessedModule,
+}
+
+/// A module processed from a path
+#[derive(Default)]
+struct ProcessedModule {
+    /// Module records of source sections, or diagnostics if parsing failed on that section.
+    ///
+    /// Modules with special extensions such as .vue could contain multiple source sections (see `PartialLoader::PartialLoader`).
+    /// Plain ts/js modules have one section. Using `SmallVec` to avoid allocations for plain modules.
+    section_module_records: SmallVec<[Result<ResolvedModuleRecord, Vec<OxcDiagnostic>>; 1]>,
```

Comment:
The term `section` is vague, what is `source sections`?

Reply:
The term `section` is taken from https://github.com/oxc-project/oxc/blob/cfdcfdb17cdded5cd0b5757ae2dec6aaf750c3f0/crates/oxc_linter/src/loader/partial_loader/mod.rs#L16. `PartialLoader` is also mentioned in the docs of `section_module_records`. 

Personally I find this term precise and couldn't come up with a better name.